### ISSUE TITLE
Detect Ctrl+Arrows

### ISSUE THREAD

**Issue by moosingin3space** [ticki/termion#46]
I ran the `keys` example and it was unable to detect any key combination including Ctrl+an Arrow key.

**Comment by osa1** [comment]
The trouble is, different terminals handle those keys differently, so unless `termion` has some kind of terminal -specific input handling this is impossible to do. To test it, use one of the programs here https://gist.github.com/osa1/86015faeb92f840f9328080dd9276bd9 with different terminals. You'll see that different byte sequences are added to `stdin` when using key combinations like `ctrl + arrow` or `alt + arrow` in different terminals.

**Comment by IGI-111** [comment]
Well I wouldn't say it's *impossible* but escape codes do surely have a way of not being standardized, even the old VT100 specs are followed only in some places.

So the spec is the implementation. Unfortunately Modifier keys coupled with special keys are not at all well supported because they are niche.

To implement that feature one would have to hunt down what the escape codes are in every family of terminal: libvte-based, xterm, rxvt, rxvt-unicode, konsole, etc and implement support for them.

There's also a bit of change in `termion::event::Key` because `Ctrl(char)`s have to become `Ctrl(Key)`s

**Comment by IGI-111** [comment]
Well after some investigation the situation is a little less bleak, although parsing is going to be quite painful.

Let's recap just for arrow keys, function keys and mouse modifiers are similar, but would require different parsing because of different codes. I just wanted to provide an example and arrow keys are among the most standardized codes for special modifiers, only xterm and rxvt syntax seems to exist for those.

Mod | Up key | Left key | Down key | Right key
------|-----------|------------|-------------|------------
None | `ESC [A` | `ESC [D` | `ESC [B` | `ESC [C`
Shift | `ESC [1;2A` <br> `ESC [a` | `ESC [1;2D` <br> `ESC [d` | `ESC [1;2B` <br> `ESC [b` | `ESC [1;2C` <br> `ESC [c`
Ctrl | `ESC [1;5A` <br> `ESC [0a` | `ESC [1;5D` <br> `ESC [0d` | `ESC [1;5B` <br> `ESC [0b` | `ESC [1;5C` <br> `ESC [0c`
Alt | `ESC [1;3A` <br> `ESC ESC [A` | `ESC [1;3D` <br> `ESC ESC [D` | `ESC [1;3B` <br> `ESC ESC [B` | `ESC [1;3C` <br> `ESC ESC [C`
Ctrl+Shift  | `ESC [1;6A` <br> `ESC [a` | `ESC [1;6D` <br> `ESC [d` | `ESC [1;6B` <br> `ESC [b` | `ESC [1;6C` <br> `ESC [c`
Ctrl+Alt | `ESC [1;7A` <br> `ESC ESC [0a` | `ESC [1;7D` <br> `ESC ESC [0d` | `ESC [1;7B` <br> `ESC ESC [0b` | `ESC [1;7C` <br> `ESC ESC [0c`
Shift+Alt | `ESC [1;4A` <br> `ESC ESC [a` | `ESC [1;4D` <br> `ESC ESC [d` | `ESC [1;4B` <br> `ESC ESC [b` | `ESC [1;4C` <br> `ESC ESC [c`
Ctrl+Alt+Shift | `ESC [1;8A` <br> `ESC ESC [a` | `ESC [1;8D` <br> `ESC ESC [d` | `ESC [1;8B` <br> `ESC ESC [b` | `ESC [1;8C` <br> `ESC ESC [c`

Note that some of these are simply not detected since some terminals bind some combos to internal functions.

At this point I feel like building a database of all known escape codes for every terminal function would be a good thing. If only it weren't so tedious to do.

Hell it's probably something you can script with a keyboard emulator.

**Comment by le-jzr** [comment]
Any progress on this? I happen to need modifier keys on more than just arrows, too.

**Comment by le-jzr** [comment]
Actually, it would be much simpler for both sides if it was possible to read events as raw data, with parsing being optional.

**Comment by gyscos** [comment]
Don't you get `Event::Unsupported(Vec<u8>)` on input that isn't recognized? This should allow custom parsing.

**Comment by le-jzr** [comment]
It's clumsy. Suppose you have custom "remember-what-I-pressed" keybindings. You don't necessarily need to know what the user pressed, just saving the byte array is enough. But the way events are encoded right now, I can't treat the inputs uniformly, I need to lower the parsed enum back into a byte array.

**Comment by gyscos** [comment]
I agree that providing raw access would be nice.
As a workaround, for the given use-case, you can save the list of all `Event`s, including parsed and non-parsed ones.

**Comment by mmstick** [comment]
Ion will also need support for this so that Liner can implement handling of C^Left and C^Right in emacs mode.

**Comment by ParadoxSpiral** [comment]
It also looks like Ctrl+Shift+char returns the same character as Ctrl+char.

**Comment by gyscos** [comment]
Some of this information is available in terminfo:

```
$ infocmp -L -1 $TERM
#	Reconstructed via infocmp from file: /usr/share/terminfo/v/vte-256color
vte-256color|VTE with xterm 256-colors,
...
	key_sr=\E[1;2A,
	key_sf=\E[1;2B,
	key_sright=\E[1;2C,
	key_sleft=\E[1;2D,
	key_shome=\E[1;2H,
	key_sprevious=\E[5;2~,
	key_snext=\E[6;2~,
...
```

`sright` is shift-right, I suppose `sr` is shift-up (reverse?) and `sf` is shift-down (forward?). `sprevious` and `snext` are shift-page-up/down.

Though this may not be 100% reliable; it's still missing ctrl/alt modifiers, and `TERM=konsole` for example does not have any of these bindings. :(

**Comment by ProgVal** [comment]
> Some of this information is available in terminfo:

Good catch! There is even a rust lib to read these: https://github.com/meh/rust-terminfo (eg. https://docs.rs/terminfo/0.5.0/terminfo/capability/struct.KeySRight.html https://docs.rs/terminfo/0.5.0/terminfo/capability/struct.KeySLeft.html )

> Though this may not be 100% reliable; it's still missing ctrl/alt modifiers, and `TERM=konsole` for example does not have any of these bindings. :(

This should probably be reported to Konsole or terminfo.
